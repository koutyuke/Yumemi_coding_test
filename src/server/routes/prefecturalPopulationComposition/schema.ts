import { t } from "elysia";

export const populationDataByYearSchema = t.Object({
  year: t.Number(),
  value: t.Number(),
});

export const prefecturalPopulationCompositionDataResponseSchema = t.Object({
  totalPopulation: t.Array(populationDataByYearSchema),
  youngPopulation: t.Array(populationDataByYearSchema),
  productivePopulation: t.Array(populationDataByYearSchema),
  elderlyPopulation: t.Array(populationDataByYearSchema),
});

export const prefecturalPopulationCompositionRouteResponseSchema = t.Object({
  boundaryYear: t.Number(),
  data: prefecturalPopulationCompositionDataResponseSchema,
});
