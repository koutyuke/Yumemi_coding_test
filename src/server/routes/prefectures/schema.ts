import { t } from "elysia";

export const prefectureDataSchema = t.Object({
  prefCode: t.Number(),
  prefName: t.String(),
});

export const prefecturesRouteResponseSchema = t.Object({
  data: t.Array(prefectureDataSchema),
});
