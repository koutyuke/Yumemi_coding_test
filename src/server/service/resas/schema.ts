import { t } from "elysia";

export const ResasAPIPrefecturePopulationResponseSchema = t.Object({
  message: t.Union([t.String(), t.Null()]),
  result: t.Object({
    boundaryYear: t.Number(),
    data: t.Array(
      t.Object({
        label: t.Union([t.Literal("総人口"), t.Literal("年少人口"), t.Literal("生産年齢人口"), t.Literal("老年人口")]),
        data: t.Array(
          t.Object({
            year: t.Number(),
            value: t.Number(),
            rate: t.Optional(t.Number()),
          }),
        ),
      }),
    ),
  }),
});

export const ResasAPIPrefecturesResponseSchema = t.Object({
  message: t.Union([t.String(), t.Null()]),
  result: t.Array(
    t.Object({
      prefCode: t.Number(),
      prefName: t.String(),
    }),
  ),
});
