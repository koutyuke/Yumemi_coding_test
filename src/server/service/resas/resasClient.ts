import { ResasAPIPrefecturePopulationResponse, ResasAPIPrefecturesResponse } from "./type";

class ResasClient {
  private baseUrl = "https://opendata.resas-portal.go.jp/api/v1";

  constructor(private readonly apiKey: string) {}

  async getPrefectures() {
    const response = await fetch(`${this.baseUrl}/prefectures`, {
      headers: {
        "X-API-KEY": this.apiKey,
      },
    });

    if (!response.ok) {
      throw new Error("Failed to fetch Resas API prefectures.");
    }

    const res = (await response.json()) as ResasAPIPrefecturesResponse;

    if (!res.result) {
      throw new Error("Response does not have result field.");
    }

    return res;
  }

  async getPrefecturePopulation(prefCode: number) {
    const response = await fetch(`${this.baseUrl}/population/composition/perYear?cityCode=-&prefCode=${prefCode}`, {
      headers: {
        "X-API-KEY": this.apiKey,
      },
    });

    if (!response.ok) {
      throw new Error("Failed to fetch Resas API prefectural population.");
    }

    const res = (await response.json()) as ResasAPIPrefecturePopulationResponse;

    if (!res.result) {
      throw new Error("Response does not have result field.");
    }

    return res;
  }
}

export { ResasClient };
