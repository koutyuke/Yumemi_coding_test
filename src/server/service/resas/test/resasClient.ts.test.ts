import { Value } from "@sinclair/typebox/value";
import { describe, test, expect } from "bun:test";
import { config } from "dotenv";
import { ResasClient } from "../resasClient";
import { ResasAPIPrefecturePopulationResponseSchema, ResasAPIPrefecturesResponseSchema } from "../schema";

config();

describe("ResasClient", () => {
  const apiKey = process.env.RESAS_API_KEY;

  if (!apiKey) {
    throw new Error("RESAS_API_KEY is not set.");
  }

  describe("getPrefectures - 都道府県一覧の取得", () => {
    test("返り値がSchemaに沿った値になっている", () => {
      expect(
        (async () => {
          const resasClient = new ResasClient(apiKey);
          const res = await resasClient.getPrefectures();
          return Value.Check(ResasAPIPrefecturesResponseSchema, res);
        })(),
      ).resolves.toBeTrue();
    });

    test("fetchに失敗した場合、エラーが発生する", () => {
      expect(
        (async () => {
          const resasClient = new ResasClient("invalid-api-key");
          await resasClient.getPrefectures();
        })(),
      ).rejects.toThrow();
    });
  });

  describe("getPrefecturePopulation - 都道府県別人口構成の取得", () => {
    test("返り値がSchemaに沿った値になっている", () => {
      expect(
        (async () => {
          const resasClient = new ResasClient(apiKey);
          const res = await resasClient.getPrefecturePopulation(11);
          return Value.Check(ResasAPIPrefecturePopulationResponseSchema, res);
        })(),
      ).resolves.toBeTrue();
    });

    test("fetchに失敗した場合、エラーが発生する", () => {
      expect(
        (async () => {
          const resasClient = new ResasClient("invalid-api-key");
          await resasClient.getPrefecturePopulation(11);
        })(),
      ).rejects.toThrow();
    });

    test("関数の返り値とSnapshotが一致する", () => {
      expect(
        (async () => {
          const resasClient = new ResasClient(apiKey);
          return resasClient.getPrefecturePopulation(11);
        })(),
      ).resolves.toMatchSnapshot();
    });
  });
});
