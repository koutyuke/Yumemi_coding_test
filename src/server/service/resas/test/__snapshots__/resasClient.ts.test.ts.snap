// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`ResasClient getPrefecturePopulation - 都道府県別人口構成の取得 関数の返り値とSnapshotが一致する 1`] = `
{
  "message": null,
  "result": {
    "boundaryYear": 2020,
    "data": [
      {
        "data": [
          {
            "value": 2430871,
            "year": 1960,
          },
          {
            "value": 3014983,
            "year": 1965,
          },
          {
            "value": 3866472,
            "year": 1970,
          },
          {
            "value": 4821340,
            "year": 1975,
          },
          {
            "value": 5420480,
            "year": 1980,
          },
          {
            "value": 5863678,
            "year": 1985,
          },
          {
            "value": 6405319,
            "year": 1990,
          },
          {
            "value": 6759311,
            "year": 1995,
          },
          {
            "value": 6938006,
            "year": 2000,
          },
          {
            "value": 7054243,
            "year": 2005,
          },
          {
            "value": 7194556,
            "year": 2010,
          },
          {
            "value": 7266534,
            "year": 2015,
          },
          {
            "value": 7344765,
            "year": 2020,
          },
          {
            "value": 7202953,
            "year": 2025,
          },
          {
            "value": 7076167,
            "year": 2030,
          },
          {
            "value": 6909319,
            "year": 2035,
          },
          {
            "value": 6721414,
            "year": 2040,
          },
          {
            "value": 6524800,
            "year": 2045,
          },
        ],
        "label": "総人口",
      },
      {
        "data": [
          {
            "rate": 30.94,
            "value": 752024,
            "year": 1960,
          },
          {
            "rate": 26.23,
            "value": 790961,
            "year": 1965,
          },
          {
            "rate": 25.94,
            "value": 1002863,
            "year": 1970,
          },
          {
            "rate": 27.8,
            "value": 1340136,
            "year": 1975,
          },
          {
            "rate": 26.91,
            "value": 1458421,
            "year": 1980,
          },
          {
            "rate": 23.54,
            "value": 1380180,
            "year": 1985,
          },
          {
            "rate": 18.69,
            "value": 1196946,
            "year": 1990,
          },
          {
            "rate": 16.13,
            "value": 1090395,
            "year": 1995,
          },
          {
            "rate": 14.77,
            "value": 1024787,
            "year": 2000,
          },
          {
            "rate": 13.98,
            "value": 986361,
            "year": 2005,
          },
          {
            "rate": 13.26,
            "value": 953668,
            "year": 2010,
          },
          {
            "rate": 12.53,
            "value": 910805,
            "year": 2015,
          },
          {
            "rate": 11.69,
            "value": 858384,
            "year": 2020,
          },
          {
            "rate": 11.37,
            "value": 819197,
            "year": 2025,
          },
          {
            "rate": 11,
            "value": 778699,
            "year": 2030,
          },
          {
            "rate": 10.77,
            "value": 744151,
            "year": 2035,
          },
          {
            "rate": 10.76,
            "value": 723391,
            "year": 2040,
          },
          {
            "rate": 10.71,
            "value": 698496,
            "year": 2045,
          },
        ],
        "label": "年少人口",
      },
      {
        "data": [
          {
            "rate": 63.59,
            "value": 1545841,
            "year": 1960,
          },
          {
            "rate": 68.54,
            "value": 2066337,
            "year": 1965,
          },
          {
            "rate": 68.93,
            "value": 2665020,
            "year": 1970,
          },
          {
            "rate": 66.85,
            "value": 3223138,
            "year": 1975,
          },
          {
            "rate": 66.87,
            "value": 3624550,
            "year": 1980,
          },
          {
            "rate": 69.25,
            "value": 4060573,
            "year": 1985,
          },
          {
            "rate": 72.76,
            "value": 4660202,
            "year": 1990,
          },
          {
            "rate": 73.63,
            "value": 4976945,
            "year": 1995,
          },
          {
            "rate": 72.23,
            "value": 5011202,
            "year": 2000,
          },
          {
            "rate": 69.35,
            "value": 4892253,
            "year": 2005,
          },
          {
            "rate": 66.01,
            "value": 4749108,
            "year": 2010,
          },
          {
            "rate": 62.03,
            "value": 4507174,
            "year": 2015,
          },
          {
            "rate": 59.02,
            "value": 4335188,
            "year": 2020,
          },
          {
            "rate": 60.39,
            "value": 4350217,
            "year": 2025,
          },
          {
            "rate": 59.6,
            "value": 4217720,
            "year": 2030,
          },
          {
            "rate": 57.93,
            "value": 4002341,
            "year": 2035,
          },
          {
            "rate": 55.05,
            "value": 3700406,
            "year": 2040,
          },
          {
            "rate": 53.51,
            "value": 3491511,
            "year": 2045,
          },
        ],
        "label": "生産年齢人口",
      },
      {
        "data": [
          {
            "rate": 5.47,
            "value": 133006,
            "year": 1960,
          },
          {
            "rate": 5.23,
            "value": 157685,
            "year": 1965,
          },
          {
            "rate": 5.14,
            "value": 198589,
            "year": 1970,
          },
          {
            "rate": 5.31,
            "value": 256014,
            "year": 1975,
          },
          {
            "rate": 6.16,
            "value": 333874,
            "year": 1980,
          },
          {
            "rate": 7.16,
            "value": 420099,
            "year": 1985,
          },
          {
            "rate": 8.28,
            "value": 530539,
            "year": 1990,
          },
          {
            "rate": 10.08,
            "value": 681172,
            "year": 1995,
          },
          {
            "rate": 12.82,
            "value": 889243,
            "year": 2000,
          },
          {
            "rate": 16.4,
            "value": 1157006,
            "year": 2005,
          },
          {
            "rate": 20.36,
            "value": 1464860,
            "year": 2010,
          },
          {
            "rate": 24.62,
            "value": 1788735,
            "year": 2015,
          },
          {
            "rate": 26.35,
            "value": 1934994,
            "year": 2020,
          },
          {
            "rate": 28.23,
            "value": 2033539,
            "year": 2025,
          },
          {
            "rate": 29.39,
            "value": 2079748,
            "year": 2030,
          },
          {
            "rate": 31.3,
            "value": 2162827,
            "year": 2035,
          },
          {
            "rate": 34.18,
            "value": 2297617,
            "year": 2040,
          },
          {
            "rate": 35.78,
            "value": 2334793,
            "year": 2045,
          },
        ],
        "label": "老年人口",
      },
    ],
  },
}
`;
